src/Data/HashMap/Linear.hs

```diff
@@ -120,141 +149,255 @@ data ProbeResult k v where
+  mapMaybeWithKey
+    (\k v -> if f k v then Just v else Nothing)
+
+-- | Complexity: O(capacity hm)
+filter :: Keyed k => (v -> Bool) -> HashMap k v #-> HashMap k v
+filter f = filterWithKey (\_k v -> f v)
+
+-- | Union of two maps using the provided function on conflicts.
+--
+-- Complexity: O(min(capacity hm1, capacity hm2)
+unionWith
+  :: Keyed k => (v -> v -> v)
+  -> HashMap k v #-> HashMap k v #-> HashMap k v
+unionWith onConflict (hm1 :: HashMap k v) hm2 =
+  -- To insert the elements in smaller map to the larger map, we
+  -- compare their capacities, and flip the arguments if necessary.
+  capacity hm1 & \(hm1', Ur cap1) ->
+    capacity hm2 & \(hm2', Ur cap2) ->
+      if cap1 > cap2
+      then go onConflict hm1' (toList hm2')
+      else go (\v2 v1 -> onConflict v1 v2) hm2' (toList hm1')
+  where
+    go :: (v -> v -> v)
+       -> HashMap k v -- ^ larger map
+       #-> Ur [(k, v)] -- ^ contents of the smaller map
+       #-> HashMap k v
+    go _ hm (Ur []) = hm
+    go f hm (Ur ((k, vr):xs)) =
+      alter hm k (\case
+        Nothing -> Just vr
+        Just vl -> Just (f vl vr))
+        & \hm -> go f hm (Ur xs)
+
+-- | A right-biased union.
+--
+-- Complexity: O(min(capacity hm1, capacity hm2)
+union :: Keyed k => HashMap k v #-> HashMap k v #-> HashMap k v
+union hm1 hm2 = unionWith (\_v1 v2 -> v2) hm1 hm2
+
+-- | Intersection of two maps with the provided combine function.
+--
+-- Complexity: O(min(capacity hm1, capacity hm2)
+intersectionWith
+  :: Keyed k
+  => (a -> b -> c)
+  -> HashMap k a #-> HashMap k b #-> HashMap k c
+intersectionWith combine (hm1 :: HashMap k a') hm2 =
+  capacity hm1 & \(hm1', Ur cap1) ->
+    capacity hm2 & \(hm2', Ur cap2) ->
+      if cap1 > cap2
+      then go combine hm1' hm2'
+      else go (\v2 v1 -> combine v1 v2) hm2' hm1'
+ where
+   go :: (a -> b -> c) -> HashMap k a #-> HashMap k b #-> HashMap k c
+   go f hm1 hm2 = Unsafe.toLinear (go' f) hm1 hm2
+
+   -- This function is using 'hm1' unsafely, because 'mapMaybeWithKey'
+   -- does not allow passing state between elements. It relies on 'lookup'
+   -- not modifying the HashMap. It is possible to implement it safely
+   -- by inlining parts of 'mapMaybeWithKey', but this one is simpler.
+   go' :: (a -> b -> c) -> HashMap k a -> HashMap k b #-> HashMap k c
+   go' f hm1 hm2 =
+     hm2
+       & mapMaybeWithKey (\k b ->
+         lookup hm1 k & \case
+           (hm', Ur Nothing) -> Unsafe.toLinear (const Nothing) hm'
+           (hm', Ur (Just a)) -> Unsafe.toLinear (const (Just (f a b))) hm'
+         )
+
+-- |
+-- Reduce the 'HashMap' 'capacity' to decrease wasted memory. Returns
+-- a semantically identical 'HashMap'.
+--
+-- This is only useful after a lot of deletes.
+--
+-- Complexity: O(capacity hm)
+shrinkToFit :: Keyed k => HashMap k a #-> HashMap k a
+shrinkToFit hm =
+  size hm & \(hm', Ur size) ->
+    let targetSize = ceiling
+          (max 1 (fromIntegral size / constMaxLoadFactor))
+    in  resize targetSize hm'
```

Comment:
The comment of resize says the `targetSize` should be larger. Just note here that the use is safe because you checked things will fit.